kaiax/valset/README.md

```diff
@@ -0,0 +1,138 @@
+- committee shuffle seed - The seed is calculated using previous block's information. The copied council is shuffled with the calculated seed to get the committee.
+
+Committee selection logic is different before/after Randao Hardfork when it's proposer policy is weightedCouncil. So the condition to activate RandaoCommittee is `policy.IsDefaultSet() || (policy.IsWeightedCouncil() && !rules.IsRandao)`
+- committee shuffle seed calculation logic
+  - before Randao: the seed is calculated using prevHash. `seed = int64(binary.BigEndian.Uint64(prevHash.Bytes()[:8]))`. 
+  - after Randao: the seed is calculated using mixHash. `seed = int64(binary.BigEndian.Uint64(mixHash[:8]))`
+- council shuffle
+  - before Randao: extract (proposer, next proposer which is differnt from proposer) and shuffle it. Attach the proposers again and slice the council.
+  - after Randao: shuffle the council and slice the committee.
+
+Example of BeforeRandaoCommittee
+```
+Condition: proposerIdx = 3, nextProposerIdx = 7, committeesize = 6, council = [0,1,2,3,4,5,6,7,8,9]
+Step1. extract proposers to committee: proposers = [3,7], council = [0,1,2,4,5,6,8,9]
+Step2. shuffle the council. proposers = [3,7], council = [4,5,6,8,9,0,1,2]
+Step3. merge. council = [3,7,4,5,6,8,9,0,1,2]
+Step4. slice the council by committee size. committee = [3,7,4,5,6,8]
+```
+### proposer: a member of committee who proposes the block
+A proposer proposes the block. We call as author after the block is created.
+#### ProposerPolicy
+Proposer is selected based on the proposer policy the network chosen. Also, each proposer selection logic has been updated per HF.
+- RoundRobin: `council(N-1)[(prevAuthorIdx+round)%len(council(N-1))]`
+- Sticky: `council(N-1)[(prevAuthorIdx+round+1)%len(council(N-1))]`
+- WeightedRandom
+  - the council is splitted into qualified, demoted addresses by minimum staking amount since istanbul HF.
+  - proposer is picked from proposers array which is created at proposerupdateinterval block with staking amount and gini. However, it is deprecated since Randao HF.
+
+The selection of proposer policy is limited by consensus algorithm.
+  - clique - 0: RoundRobin [default: 0]
+  - istanbul - 0: RoundRobin, 1: Sticky, 2: WeightedRandom [default: 0]
+
+## Persistent Schema
+The voting blks and the council addressList is stored at miscDB
+### Voting Blks
+- ReadVoteBlks - it reads whole addvalidator/removevalidator voting blks 
+- StoreVoteBlks - it stores/updates whole addvalidator/removevalidator voting blks
+
+### ValSetSnapshot
+- ReadValSetSnapshot - it reads the closest ValSetSnapshot of block N. 
+- StoreValSetSnapshot - deprecated. kaiax no longer stores ValSetSnapshot.
+
+### Valset
+- ReadCouncilAddressListFromDb(n) - n is the addvalidator/removevalidator voting blks
+- WriteCouncilAddressListToDb(n, council) - n is the addvalidator/removevalidator voting blks
+
+## In-memory Structures
+###  Council
```

Comment:
#### 1. Council, CouncilContext 
```
type Council []address
type CouncilContext {
	Header
}
```

The only context is `rules`, so CouncilContext structure is not needed.

Additionally, the council structure is changed like below.
```
type council struct {
	blockNumber uint64
	rules       params.Rules

	qualifiedValidators valset.AddressList // qualified is a list of validators which will be the member of the committee
	demotedValidators   valset.AddressList // demoted are subtract of qualified from council(N)

	councilAddressList valset.AddressList // total council node address list. the order is reserved.
}
```
The change was made because the sources for the committee, proposer, and API calls differ. To streamline the process, all relevant groups—qualified, demoted, and council members—are included upfront, with qualification filtering applied in advance.

#### 2. Committee, CommitteeContext
```
type Committee []address
type CommitteeContext {
	Council
	StakingInfo
	Config
        Proposers
}
```
I'll consider this structure. Creating a separate Committee type doesn’t seem necessary, but a committeeContext seems needed and possible. Missing from the suggested structure are elements like `proposers` and `prevAuthor`. But I'll impl it after opening this PR from drafted.

#### 3. getter structure
```
ValsetModule.getCouncilContext(N) -> CouncilContext
CouncilContext.GetCouncil(N) -> Council
```
-> CouncilContext doesn't seem necessary so ValsetModule.GetCouncil(N) is enough.

```
CommitteeContext.GetCommittee(N, R) -> Committee
CommitteeContext.GetProposer(N, R) -> Proposer
```
-> I'll impl it.
```
Council.IsMember(addr) -> bool
Committee.IsMember(addr) -> bool
```
-> IsMember is implemented as valSet.AddressList().GetIdxByAddress